File: mcp-code-intelligence-kotlin/src/main/kotlin/com/codeintel/parsers/GrammarRegistry.kt
/**
 * KSA-172: Grammar Registry — Manages tree-sitter grammar loading and caching.
 */
package com.codeintel.parsers

data class LanguageConfig(
    val id: String,
    val extensions: List<String>,
    val grammarPath: String,
    val parserClass: String,
)

data class GrammarRegistryConfig(
    val languages: List<LanguageConfig> = emptyList(),
    val grammarDir: String = "",
)

class GrammarRegistry(private val config: GrammarRegistryConfig) {
    private val languageParsers = mutableMapOf<String, ILanguageParser>()
    private val extensionMap = mutableMapOf<String, String>()
    private val unavailable = mutableSetOf<String>()
    private var initialized = false

    init { buildExtensionMap() }

    fun initialize() { if (initialized) return; initialized = true }

    fun getParser(filePath: String): ILanguageParser? {
        if (!initialized) initialize()
        val ext = ".${filePath.substringAfterLast('.', "")}"
        val langId = extensionMap[ext] ?: return null
        if (langId in unavailable) return null
        return languageParsers[langId] ?: loadParser(langId)
    }

    fun getLanguageId(filePath: String): String? {
        val ext = ".${filePath.substringAfterLast('.', "")}"
        return extensionMap[ext]
    }

    fun registerParser(langId: String, parser: ILanguageParser) {
        languageParsers[langId] = parser
    }

    fun isAvailable(langId: String): Boolean =
        langId !in unavailable && config.languages.any { it.id == langId }

    private fun loadParser(langId: String): ILanguageParser? {
        val langConfig = config.languages.find { it.id == langId } ?: return null
        return try {
            val cls = Class.forName(langConfig.parserClass)
            val parser = cls.getDeclaredConstructor(String::class.java).newInstance(langId) as ILanguageParser
            languageParsers[langId] = parser
            parser
        } catch (e: Exception) {
            System.err.println("[grammar-registry] Failed to load $langId: ${e.message}")
            unavailable.add(langId)
            null
        }
    }

    private fun buildExtensionMap() {
        for (lang in config.languages) {
            for (ext in lang.extensions) { extensionMap[ext] = lang.id }
        }
    }
}

File: mcp-code-intelligence-python/src/mcp_code_intel/parsers/grammar_registry.py
"""
KSA-178: Grammar Registry — Manages tree-sitter grammar loading and caching.
Maps file extensions to language parsers, lazy-loads grammars on first use.
Port of mcp-code-intelligence-nodejs/src/parsers/grammar-registry.ts.
"""

from __future__ import annotations

import importlib
import os
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any

from tree_sitter import Language, Parser

from .types import ILanguageParser


@dataclass
class LanguageConfig:
    id: str
    extensions: list[str]
    grammar_path: str  # Path to .so/.dll shared library
    parser_module: str  # Python module path for the language parser class


@dataclass
class GrammarRegistryConfig:
    languages: list[LanguageConfig] = field(default_factory=list)
    grammar_dir: str = ""


class GrammarRegistry:
    """Manages tree-sitter grammar loading and language parser instantiation."""

    def __init__(self, config: GrammarRegistryConfig) -> None:
        self._config = config
        self._parsers: dict[str, Parser] = {}
        self._language_parsers: dict[str, ILanguageParser] = {}
        self._extension_map: dict[str, str] = {}
        self._unavailable: set[str] = set()
        self._initialized = False
        self._build_extension_map()

    def initialize(self) -> None:
        """Initialize tree-sitter runtime. Must be called before parsing."""
        if self._initialized:
            return
        self._initialized = True

    def get_parser(self, file_path: str) -> ILanguageParser | None:
        """Get a parser for a file path based on extension. Returns None if unsupported."""
        if not self._initialized:
            self.initialize()

        ext = os.path.splitext(file_path)[1].lower()
        lang_id = self._extension_map.get(ext)

        if not lang_id or lang_id in self._unavailable:
            return None

        if lang_id in self._language_parsers:
            return self._language_parsers[lang_id]

        return self._load_parser(lang_id)

    def get_language_id(self, file_path: str) -> str | None:
        """Get language ID for a file extension."""
        ext = os.path.splitext(file_path)[1].lower()
        return self._extension_map.get(ext)

    def list_languages(self) -> list[dict[str, Any]]:
        """List all registered languages."""
        return [
            {
                "id": lang.id,
                "extensions": lang.extensions,
                "available": lang.id not in self._unavailable,
            }
            for lang in self._config.languages
        ]

    def is_available(self, lang_id: str) -> bool:
        """Check if a language grammar is available."""
        return (
            lang_id not in self._unavailable
            and any(l.id == lang_id for l in self._config.languages)
        )

    def _load_parser(self, lang_id: str) -> ILanguageParser | None:
        lang_config = next(
            (c for c in self._config.languages if c.id == lang_id), None
        )
        if not lang_config:
            return None

        try:
            grammar_path = os.path.join(self._config.grammar_dir, lang_config.grammar_path)

            if not os.path.exists(grammar_path):
                self._unavailable.add(lang_id)
                return None

            language = Language(grammar_path, lang_id)
            parser = Parser()
            parser.set_language(language)
            self._parsers[lang_id] = parser

            # Dynamically import the language parser module
            module = importlib.import_module(lang_config.parser_module)
            parser_class = module.Parser  # Each module exports a Parser class
            lang_parser: ILanguageParser = parser_class(parser, lang_id)
            self._language_parsers[lang_id] = lang_parser

            return lang_parser
        except Exception as e:
            import sys
            print(f"[grammar-registry] Failed to load {lang_id}: {e}", file=sys.stderr)
            self._unavailable.add(lang_id)
            return None

    def _build_extension_map(self) -> None:
        for lang in self._config.languages:
            for ext in lang.extensions:
                self._extension_map[ext] = lang.id


def load_grammar_config(config_path: str) -> GrammarRegistryConfig:
    """Load grammar registry config from JSON file."""
    import json

    with open(config_path, "r") as f:
        data = json.load(f)

    languages = [
        LanguageConfig(
            id=lang["id"],
            extensions=lang["extensions"],
            grammar_path=lang.get("grammarPath", ""),
            parser_module=lang.get("parserModule", ""),
        )
        for lang in data.get("languages", [])
    ]

    return GrammarRegistryConfig(
        languages=languages,
        grammar_dir=str(Path(config_path).parent),
    )
